src/solve/coinduction.md

```diff
@@ -0,0 +1,201 @@
+# Coinduction
+
+The trait solver may use coinduction when proving goals. Coinduction is fairly subtle so we're giving it its own chapter.
+
+## Coinduction and induction
+
+With induction, we recursively apply proofs until we end up with a finite proof tree. Consider the example of `Vec<Vec<Vec<u32>>>: Debug` which results in the following tree.
+
+- `Vec<Vec<Vec<u32>>>: Debug`
+    - `Vec<Vec<u32>>: Debug`
+        - `Vec<u32>: Debug`
+            - `u32: Debug`
+
+This tree is finite. But not all goals we would want to hold have finite proof trees, consider the following example:
+
+```rust
+struct List<T> {
+    value: T,
+    next: Option<Box<List<T>>>,
+}
+```
+
+For `List<T>: Send` to hold all its fields have to recursively implement `Send` as well. This would result in the following proof tree:
+
+- `List<T>: Send`
+    - `T: Send`
+    - `Option<Box<List<T>>>: Send`
+        - `Box<List<T>>: Send`
+            - `List<T>: Send`
+                - `T: Send`
+                - `Option<Box<List<T>>>: Send`
+                    - `Box<List<T>>: Send`
+                        - ...
+
+This tree would be infinitely large which is exactly what coinduction is about. 
+
+> To **inductively** prove a goal you need to provide a finite proof tree for it.
+> To **coinductively** prove a goal the provided proof tree may be infinite.
+
+## Why is coinduction correct
+
+When checking whether some trait goals holds, we're asking "does there exist an `impl`
+which satisfies this bound". Even if are infinite chains of nested goals, we still have a
+unique `impl` which should be used.
+
+## How to implement coinduction
+
+While our implementation can not check for coninduction by trying to construct an infinite
+tree as that would take infinite ressources, it still makes sense to think of coinduction
+from this perspective.
+
+As we cannot check for infinite trees, we instead search for patterns for which we know that
+they would result in an infinite proof tree. The currently pattern we detect are (canonical)
+cycles. If `T: Send` relies on `T: Send` then it's pretty clear that this will just go on forever.
+
+With cycles we have to be careful with caching. Due to canonicalization of regions and inference
+variables we also have to rerun queries until the provisional result returned when hitting the cycle
```

Comment:
I don't think you mentioned "provisional result" yet at this point so i expect nobody will know what this means. It also might be worth linking to https://hackmd.io/M5vJyLMGSTaeY0n96wqEWQ somewhere or inlining it since it steps through the "rerun queries until the provisional result" stuff.